src/algebra/continued_fractions/definitions/basic.lean

```diff
@@ -0,0 +1,285 @@
+- https://en.wikipedia.org/wiki/Generalized_continued_fraction
+- [Wall, H.S., *Analytic Theory of Continued Fractions*][wall2018analytic]
+
+## Tags
+
+numerics, number theory, approximations, fractions
+-/
+
+
+/-!
+A *generalised continued fraction* (gcf) is a potentially infinite expression of the form
+
+                                a₀
+                h + --------------------------
+                                  a₁
+                      b₀ + --------------------
+                                    a₂
+                            b₁ + --------------
+                                        a₃
+                                  b₂ + --------
+                                      b₃ + ...
+
+where `h` is called the *head term* or *integer part*, the `aᵢ` are called the
+*partial numerators* and the `bᵢ` the *partial denominators* of the gcf. For convenience, one often
+writes `[h; (a₀, b₀), (a₁, b₁), (a₂, b₂),...]`.
+-/
+
+-- Fix a carrier `α`.
+variable (α : Type*)
+
+/-- We collect a partial numerator `aᵢ` and partial denominator `bᵢ` in a pair `⟨aᵢ,bᵢ⟩`. -/
+protected structure gcf.pair := (a : α) (b : α)
+
+/- Interlude: define some expected coercions and instances. -/
+namespace gcf.pair
+
+/-- Make a gcf.pair printable. -/
+instance [has_repr α] : has_repr (gcf.pair α) :=
+⟨λ p, "(a : " ++ (repr p.a) ++ ", b : " ++ (repr p.b) ++ ")"⟩
+
+section coe
+/-- Fix another type `β` and assume `α` can be converted to `β`. -/
+variables {α} {β : Type*} [has_coe α β]
+
+/-- Coerce a gcf.pair by elementwise coercion. -/
+instance has_coe_to_gcf_pair : has_coe (gcf.pair α) (gcf.pair β) := ⟨λ ⟨a, b⟩, ⟨(a : β), (b : β)⟩⟩
+
+@[simp, move_cast]
+lemma coe_to_gcf_pair {a b : α} : (↑(gcf.pair.mk a b) : gcf.pair β) = gcf.pair.mk (a : β) (b : β) :=
+rfl
+
+end coe
+end gcf.pair
+
+/-- A gcf consists of a head term `h` and a sequence of gcf.pairs `s`. -/
+structure gcf := (h : α) (s : seq $ gcf.pair α)
+
+variable {α}
+
+namespace gcf
+
+/-- Returns the sequence of partial numerators `aᵢ` of `g`. -/
+def partial_numerators (g : gcf α) : seq α := g.s.map gcf.pair.a
+/-- Returns the sequence of partial denominators `bᵢ` of `g`. -/
+def partial_denominators (g : gcf α) : seq α := g.s.map gcf.pair.b
+
+/-- A gcf terminates at position `n` if its sequence terminates at position `n`. -/
+def terminates_at (g : gcf α) (n : ℕ) : Prop := g.s.terminates_at n
+
+/-- It is decidable whether a gcf terminates at a given position. -/
+instance terminates_at_decidable (g : gcf α) (n : ℕ) : decidable (g.terminates_at n) :=
+by { unfold terminates_at, apply_instance }
+
+/-- A gcf terminates if its sequence terminates. -/
+def terminates (g : gcf α) : Prop := g.s.terminates
+
+/- Interlude: define some expected coercions. -/
+section coe
+/-- Fix another type `β` and assume `α` can be converted to `β`. -/
+variables {β : Type*} [has_coe α β]
+
+/-- Coerce a sequence by elementwise coercion. -/
+def seq.coe_to_seq : has_coe (seq α) (seq β) := ⟨seq.map (λ a, (a : β))⟩
```

Comment:
I was not sure if this should be put into the sequence library and made an instance by default, so I kept it here and made it a local instance. Please let me know your opinion.

Reply:
Yes, let's keep this local. We don't, for example, have global instances letting you do this for `list`.